packages/wrangler/src/d1/migrations.tsx

```diff
@@ -0,0 +1,421 @@
+		}
+
+		for (const migration of unappliedMigrations) {
+			let query = fs.readFileSync(
+				`${migrationsPath}/${migration.Name}`,
+				"utf8"
+			);
+			query += `
+								INSERT INTO ${database.migrationsTableName} (name)
+								values ('${migration.Name}');
+						`;
+
+			let success = true;
+			let errorNotes: Array<string> = [];
+			try {
+				const response = await executeSql(
+					local,
+					config,
+					name,
+					undefined,
+					persistTo,
+					undefined,
+					query
+				);
+
+				if (response === null) {
+					// TODO:  return error
+					return;
+				}
+
+				for (const result of response) {
+					// When executing more than 1 statement, response turns into an array of QueryResult
+					if (Array.isArray(result)) {
+						for (const subResult of result) {
+							if (!subResult.success) {
+								success = false;
+							}
+						}
+					} else {
+						if (!result.success) {
+							success = false;
+						}
+					}
+				}
+			} catch (e) {
+				const err = e as ParseError;
+
+				success = false;
+				errorNotes = err.notes.map((msg) => msg.text);
+			}
+
+			migration.Status = success ? "✅" : "❌";
+
+			render(
+				<Box flexDirection="column">
+					<Table
+						data={unappliedMigrations}
+						columns={["Name", "Status"]}
+					></Table>
+					{errorNotes.length > 0 && (
+						<Box flexDirection="column">
+							<Text>&nbsp;</Text>
+							<Text>
+								❌ Migration {migration.Name} failed with following Errors
+							</Text>
+							<Table
+								data={errorNotes.map((err) => {
+									return { Error: err };
+								})}
+							></Table>
+						</Box>
+					)}
+				</Box>
+			);
+
+			if (errorNotes.length > 0) return;
+		}
+	}
+);
+
+export const listAppliedMigrations = async (
+	migrationsTableName: string,
+	local: undefined | boolean,
+	config: ConfigFields<DevConfig> & Environment,
+	name: string,
+	persistTo: undefined | string
+): Promise<Migration[]> => {
+	const Query = `SELECT *
+									 FROM ${migrationsTableName}
+									 ORDER BY id`;
+
+	const response: QueryResult[] | null = await executeSql(
+		local,
+		config,
+		name,
+		undefined,
+		persistTo,
+		undefined,
+		Query
+	);
+
+	if (!response || response[0].results.length === 0) return [];
+
+	return response[0].results as Migration[];
+};
+
+const initMigrationsTable = async (
+	migrationsTableName: string,
+	local: undefined | boolean,
+	config: ConfigFields<DevConfig> & Environment,
+	name: string,
+	persistTo: undefined | string
+) => {
+	return executeSql(
+		local,
+		config,
+		name,
+		undefined,
+		persistTo,
+		undefined,
+		`
+						CREATE TABLE IF NOT EXISTS ${migrationsTableName}
+						(
+								id         INTEGER PRIMARY KEY AUTOINCREMENT,
+								name       TEXT UNIQUE,
+								applied_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP NOT NULL
+						);
+				`
+	);
+};
+
+type MigrationsCreateArgs = { config?: string; name: string; message: string };
+
+export function CreateOptions(yargs: Argv): Argv<MigrationsCreateArgs> {
```

Comment:
why is this so far above the handler? can we keep them together

Reply:
Fixed